### PR TITLE
Workspace file fs improvement

### PR DESCRIPTION
**Pull request by Sithi5**

### CODE REVIEW

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 1]
```diff
@@ -338,7 +338,17 @@ async def read_bytes(self, path: AnyPath, size: int = -1, offset: int = 0) -> by
 
     async def open_file(self, path: AnyPath, mode="r"):
         path = FsPath(path)
-        _, fd = await self.transactions.file_open(path, mode)
+        # check if file exist, if not and a writting mode is set, then create the file
+        if "x" in mode:
+            try:
+                _, fd = await self.transactions.file_create(path, open=True)
+            except FileExistsError:
+                raise
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py:346]: No need for a try/except here :wink:

File: parsec/core/gui/files_widget.py [hunk 2]
```diff
@@ -701,9 +697,12 @@ def _on_delete_error(self, job):
...
     def _on_folder_stat_success(self, job):
-        self.current_directory, self.current_directory_uuid, files_stats, default_selection = (
-            job.ret
-        )
+        (
+            self.current_directory,
+            self.current_directory_uuid,
+            files_stats,
+            default_selection,
+        ) = job.ret
```
[the remark this answers is not in the document]
Comment (Sithi5) [reply parsec/core/gui/files_widget.py]: Yes ! Fixed

File: tests/core/fs/workspacefs/test_workspacefile.py [hunk 3]
```diff
@@ -253,6 +290,57 @@ async def test_read_write(alice_workspace, trio_file, random_text):
...
+    # Opening with rb+
+    f = await alice_workspace.open_file("/foo/bar", "rb+")
+    triof = await trio.open_file(trio_file, "rb+")
+    # test write
+    await write_in_both_files(triof, f, new_text)
+    assert f.tell() == await triof.tell()
+    # test read
+    await compare_read(triof, f)
+    await f.close()
+    await triof.aclose()
```
[the remark this answers is not in the document]
Comment (Sithi5) [reply tests/core/fs/workspacefs/test_workspacefile.py:328]: Done

File: parsec/core/fs/workspacefs/workspacefile.py [hunk 4]
```diff
@@ -90,6 +111,10 @@ async def close(self):
             self._state = FileState.CLOSED
             await self._transactions.fd_close(self._fd)
 
+    async def aclose(self):
+        """Same as close"""
+        return await self.close()
```
[the remark this answers is not in the document]
Comment (Sithi5) [reply parsec/core/fs/workspacefs/workspacefile.py]: Done, I removed the `aclose()` method

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 1 above)
Comment (Sithi5) [reply parsec/core/fs/workspacefs/workspacefs.py]: Fixed :)

File: parsec/core/fs/workspacefs/workspacefile.py [hunk 5]
```diff
@@ -47,19 +49,22 @@ class WorkspaceFile:
...
-        if sum(c in mode for c in "rwa") != 1:
+        if sum(c in mode for c in "rwax") != 1:
             raise ValueError("must have exactly one of create/read/write/append mode")
         # Preventing to open with non-existant mode
-        elif re.search("[^arwb]", mode) is not None:
-            raise ValueError("invalid mode: ", mode)
+        elif re.search("[^arwxb+]", mode) is not None:
+            raise ValueError(f"invalid mode: '{mode}'")
+        if "b" not in mode:
+            raise NotImplementedError("Must use binary mode")
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefile.py:67]: Maybe `"Text mode is not supported at the moment"`?

Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefile.py:67]: Maybe `"Text mode is not supported at the moment"`?

File: parsec/core/fs/workspacefs/workspacefile.py [hunk 6]
```diff
@@ -73,13 +78,29 @@ async def __aexit__(self, *args):
     async def ainit(self):
         """Initializing the File Object.
 
-            Check the FileState and truncate the file if needed.
+            Check the FileState, open or create the file, truncate it and move the offset if needed.
         """
         if self._state == FileState.INIT:
             self._state = FileState.OPEN
-            # Checking if the 'w' open mode have been selected, if yes, truncate the file
+            # opening or creating the file
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefile.py:85]: Comments are capitalized: https://www.python.org/dev/peps/pep-0008/#comments

File: parsec/core/fs/workspacefs/workspacefile.py (hunk 5 above)
Comment (Sithi5) [reply parsec/core/fs/workspacefs/workspacefile.py:67]: Done 👍

File: parsec/core/fs/workspacefs/workspacefile.py (hunk 6 above)
Comment (Sithi5) [reply parsec/core/fs/workspacefs/workspacefile.py:85]: Done

**Pull request merged**